### ISSUE TITLE
etcdmain: database file does not match with snapshot

### ISSUE THREAD

**Issue by chaithramohan** [coreos/etcd#9021]
Hello,

We use a kubernetes cluster with 3 or 2 etcds and we see that sometime one of the etcd member goes into the below issue:

we see the issue that the etcd can't speak with his cluster members anymore. because his data is inconsistent or differs from the other nodes in the cluster.
You see following error message
2017-12-01 10:10:01.168631 I | etcdserver: recovered store from snapshot at index 3426537
2017-12-01 10:10:01.168656 I | etcdserver: name = etcd1
2017-12-01 10:10:01.168660 I | etcdserver: data dir = /var/lib/etcd/data
2017-12-01 10:10:01.168666 I | etcdserver: member dir = /var/lib/etcd/data/member
2017-12-01 10:10:01.168670 I | etcdserver: dedicated WAL dir = /var/lib/etcd/wal
2017-12-01 10:10:01.168673 I | etcdserver: heartbeat = 100ms
2017-12-01 10:10:01.168676 I | etcdserver: election = 1000ms
2017-12-01 10:10:01.168679 I | etcdserver: snapshot count = 10000
2017-12-01 10:10:01.168690 I | etcdserver: advertise client URLs = http://10.131.134.18:2379
2017-12-01 10:10:01.168789 W | wal: ignored file 0000000000000075-00000000003012ad.wal.broken in wal
2017-12-01 10:10:01.168799 W | wal: ignored file 0000000000000076-0000000000311951.wal.broken in wal
2017-12-01 10:10:01.168806 W | wal: ignored file 0000000000000077-0000000000320994.wal.broken in wal
2017-12-01 10:10:01.168813 W | wal: ignored file 0000000000000078-0000000000330637.wal.broken in wal
2017-12-01 10:10:01.522668 I | etcdserver: restarting member b7ac64605e023502 in cluster 8919103582a07674 at commit index 3445035
017-12-01 10:10:01.527818 C | raft: b7ac64605e023502 state.commit 3445035 is out of range [3426537, 3428606]
panic: b7ac64605e023502 state.commit 3445035 is out of range [3426537, 3428606]

We can see in the ectdctl logs the below message
etcdmain: database file (/var/lib/etcd/data/member/snap/db index 137632097) does not match with snapshot (index 137648093)

What we do as work around is to disable the puppet agent there, change the initial state in etcd.yaml from new to existing and then clean up the data folder.

Then the etcd pod that shows the issue is deleted and we find the node is back with no issues.

Please let me know if this is fixed in any upcoming versions or any other solution to find a better solution ?

The version we use is:
etcdctl --version
etcdctl version: 3.1.10
API version: 2

curl -L http://127.0.0.1:2379/version
{"etcdserver":"3.2.11","etcdcluster":"3.2.0"}

Kubernetes version : v1.5.7

**Comment by chaithramohan** [comment]
Hello,

We use a kubernetes cluster with 3 or 2 etcds and we see that sometime one of the etcd member goes into the below issue:

we see the issue that the etcd can't speak with his cluster members anymore. because his data is inconsistent or differs from the other nodes in the cluster.
You see following error message
2017-12-01 10:10:01.168631 I | etcdserver: recovered store from snapshot at index 3426537
2017-12-01 10:10:01.168656 I | etcdserver: name = etcd1
2017-12-01 10:10:01.168660 I | etcdserver: data dir = /var/lib/etcd/data
2017-12-01 10:10:01.168666 I | etcdserver: member dir = /var/lib/etcd/data/member
2017-12-01 10:10:01.168670 I | etcdserver: dedicated WAL dir = /var/lib/etcd/wal
2017-12-01 10:10:01.168673 I | etcdserver: heartbeat = 100ms
2017-12-01 10:10:01.168676 I | etcdserver: election = 1000ms
2017-12-01 10:10:01.168679 I | etcdserver: snapshot count = 10000
2017-12-01 10:10:01.168690 I | etcdserver: advertise client URLs = http://10.131.134.18:2379
2017-12-01 10:10:01.168789 W | wal: ignored file 0000000000000075-00000000003012ad.wal.broken in wal
2017-12-01 10:10:01.168799 W | wal: ignored file 0000000000000076-0000000000311951.wal.broken in wal
2017-12-01 10:10:01.168806 W | wal: ignored file 0000000000000077-0000000000320994.wal.broken in wal
2017-12-01 10:10:01.168813 W | wal: ignored file 0000000000000078-0000000000330637.wal.broken in wal
2017-12-01 10:10:01.522668 I | etcdserver: restarting member b7ac64605e023502 in cluster 8919103582a07674 at commit index 3445035
017-12-01 10:10:01.527818 C | raft: b7ac64605e023502 state.commit 3445035 is out of range [3426537, 3428606]
panic: b7ac64605e023502 state.commit 3445035 is out of range [3426537, 3428606]

We can see in the ectdctl logs the below message
etcdmain: database file (/var/lib/etcd/data/member/snap/db index 137632097) does not match with snapshot (index 137648093)

What we do as work around is to disable the puppet agent there, change the initial state in etcd.yaml from new to existing and then clean up the data folder.

Then the etcd pod that shows the issue is deleted and we find the node is back with no issues.

Please let me know if this is fixed in any upcoming versions or any other solution to find a better solution ?

The version we use is:
etcdctl --version
etcdctl version: 3.1.10
API version: 2

Kubernetes version : v1.5.7

**Comment by hexfusion** [comment]
Hi @chaithramohan would you mind updating your title to something a little more meaningful so others can reference. Maybe `etcdmain: database file does not match with snapshot` Also, you could consolidate these two messages here. Thank you!

**Comment by chaithramohan** [comment]
@hexfusion done

**Comment by xiang90** [comment]
it is probably caused by the broken wal files. do you know how the wal files got broken at the first place? did the machine experience any disk issues?

**Comment by xiang90** [comment]
Also are you able to reproduce this problem? If not, are you willing to share your broken data dir for us to investigate?

**Comment by chaithramohan** [comment]
Not able to reproduce the problem, did not face the issue yet again after that to share the data file as well but wanted to know if there is any fix in any upcoming release or is there a way to automatically handle such a failure ??

**Comment by xiang90** [comment]
I am not sure what is the root cause, so I cannot provide you the answer.

**Comment by chaithramohan** [comment]
okay, i will try getting more details once i get hold of them. In the meanwhile with regard to a similar issue (https://github.com/coreos/etcd/issues/8935), some investigation was going on in 

"I will be surprised that etcd 3.1.11 fixes the root cause. I feel the root cause is somewhere in raft pkg, and it is a pretty subtle one.

We probably should still spend sometime to understand why it happens under 3.0.10, and if 3.1.x fixes it which commit exactly fixed it "

This problem had occured when we used v3.1.5 and now its like about 2 days we upgraded to 3.2.11. Is there a possibility with regard to the above comment that 3.1.11 fixes it or any investigation in that regard ?
